Equals function for Vector with random sample

jaewookjsuh opened the pull request:
**Equals function for Vector**
- `parameter.py`: Added `get_names` function, which collects all names contained in a `Parameter` object
- `vector.py`: Implemented `equals_random_sample` function for `Vector`
  - Determines whether two `Vector` objects are equal by comparing numerical values obtained by substituting all parameters with randomly generated numbers.
  - To do so, implemented `get_names`, which collects all names contained in `Parameter` objects used for a `Vector` object
- `vector_test.py`: Added a related test.

jaewookjsuh committed:
Equals function for Vector with random sample

- Implemented `equals_random_sample` function for `Vector`
- Added a related test

jaewookjsuh committed:
Merge branch 'main' into b/sjw/equals_random

jaewookjsuh committed:
fix `get_names` for `DictRepresentation`

`parameter.py`
- add support for `ParameterByDictRepresentation` in `get_names`
- add corresponding unit test

`vector_test.py`
- add more unit tests in `test_equals_random_sample`

jaewookjsuh committed:
minor: fix `get_names` type hints

jaewookjsuh committed:
test: relax vector large-scale timing threshold

BichengYing commented on pepflow/vector.py:
```diff
@@ -671,3 +672,41 @@ def repr_by_basis(
+    def equals_random_sample(
+        self,
+        other,
+        ctx: pc.PEPContext | None = None,
```
add a seed that allow to control the randomized. The default it is still None so that it is changing ever time. Also, add repetition.  (I think default to be 3). If two expressions are equal, no matter how many different seeds, it always give the same value. If not, the "false positive" probably will be decreased by exponential of repition.

jaewookjsuh replied:
Thank you very much for the suggestion. I addressed this point.

BichengYing commented on pepflow/vector.py:
```diff
@@ -671,3 +672,41 @@ def repr_by_basis(
+
+        em = ExpressionManager(ctx, resolve_parameters=resolve_parameters_random)
+        diff_resolved = em.eval_vector(diff, sympy_mode=False).coords
+        return not np.any(np.abs(diff_resolved) >= 1e-12)
```
1e-12 might be too small. To equivalent expression with different operation order indeed can make difference a large than this. Try 10e-9 or bigger (Repetition will be more save you on this)

jaewookjsuh replied:
I revised it. Thank you!

BichengYing commented on pepflow/vector.py:
```diff
@@ -671,3 +672,41 @@ def repr_by_basis(
         em = ExpressionManager(ctx, resolve_parameters=resolve_parameters)
         return em.repr_vector_by_basis(self, sympy_mode=sympy_mode)
+
+    def get_names(self) -> set[str]:
```
how about `get_param_names`?

jaewookjsuh replied:
Yes, I think `get_param_names` is clearer. I updated it. Thanks.

BichengYing commented on pepflow/vector.py:
```diff
@@ -671,3 +672,41 @@ def repr_by_basis(
+    def get_names(self) -> set[str]:
+        """A function that collects all `name`s in a :class:`Vector` object."""
+        names = set()
+        if isinstance(self.eval_expression, VectorRepresentation):
```
what about `VectorByBasisRepresentation` case?

jaewookjsuh replied:
Good point, thanks. I added this case.

BichengYing replied:
usually avoid this missing case, we will add a else block and raise value for some unexpected cases

BichengYing commented on pepflow/vector.py:
```diff
@@ -671,3 +672,41 @@ def repr_by_basis(
+                names.update(right_vector.get_names())
+        return names
+
+    def equals_random_sample(
```
this name sounds ambiguous to me. How about `equiv_by_randomness`?

jaewookjsuh replied:
Sounds good. I changed it.

jaewookjsuh committed:
Address randomized `Vector` equivalence feedback

- Rename `equals_random_sample` to `equiv_by_randomness`
- Rename `get_names` to `get_param_names` for parameter-name collection
- Add `seed`, `repetitions`, and `atol` options
- Collect parameter names from `VectorByBasisRepresentation` coefficients
- Add tests for seeded repetition, tolerance validation, and basis representations

BichengYing commented on pepflow/vector.py:
```diff
@@ -671,3 +672,56 @@ def repr_by_basis(
+        ctx: pc.PEPContext | None = None,
+        seed: int | None = None,
+        repetitions: int = 3,
+        atol: float = 1e-9,
```
Can you double check that 1e-9 will not cause false alarm (it is equiv but diff due to numerical error)? i feel to make it little bit more loose like 1e-7. It is unlikely to have violation after 3 or more repetitions i feel.

BichengYing reviewed (approved):
LGTM!
